main/java/us/klette/constantstring/CString.java

```diff
@@ -0,0 +1,143 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package us.klette.constantstring;
+
+import javax.annotation.Nonnull;
+
+/**
+ * The CString interface defines the allowed operations
+ * within ConstantString.
+ * <p>
+ * This is a small utility library for doing constant time string manipulation.
+ * Or in simpler terms, you can do operations on strings that are done in a
+ * fixed time, and delay the linear time computation until you evaluate your
+ * set of operations.
+ *
+ * @author Kristian Klette (klette@klette.us)
+ */
+public interface CString {
+
+    /**
+     * Creates a new CString instance from a String.
+     *
+     * @param value The String to use as a base value
+     * @return The CString instance from the string
+     */
+    static CString create(final String value) {
+        return new CStringLeaf(value);
+    }
+
+    /**
+     * Evaluates the CString into a String.
+     *
+     * @return The resulting {@link String} instance.
+     */
+    @Override
+    String toString();
+
+    /**
+     * Applies a substring operations similar to
+     * {@link java.lang.String#substring(int)}.
+     *
+     * @param index The index of which to start the new string
+     * @return The CString representation after the applied substring operation
+     */
+    default CString substring(final int index) {
+        return new CStringSub(this, index);
+    }
+    /**
+     * Applies a substring operations similar to
+     * {@link java.lang.String#substring(int, int)}.
+     *
+     * @param index The index of which to start the new string
+     * @param end The end index of the substring
+     * @return The CString representation after the applied substring operation
+     */
+    default CString substring(final int index, final int end) {
+        return new CStringSubRange(this, index, end);
+    }
+
+    /**
+     * Append another CString after the current instance.
+     *
+     * @param other The instance to be appended.
+     * @return The CString representation of the two
+     */
+    default CString concat(@Nonnull final CString other) {
+        return new CStringConcat(this, other);
+    }
+    /**
+     * Append another String after the current instance.
+     *
+     * @param other The instance to be appended.
+     * @return The CString representation of the two
+     */
+    default CString concat(@Nonnull final String other) {
+        return this.concat(new CStringLeaf(other));
+    }
+
+    /**
+     * Delete a subsection of the String.
+     *
+     * @param start The start index of the delete
+     * @param end The end index of the deleted part
+     * @return The CString representation with the delete applied.
+     */
+    default CString delete(final int start, final int end) {
+        return new CStringConcat(
+                new CStringSubRange(this, 0, start),
+                new CStringSub(this, start+end)
```

Comment:
hm, why isn't it just `concat(substring(0, start), substring(end))`? (not sure about inclusive/enclusive and thus +-1)
I.e. 

1. why the Sub is for `start+end`?!  If I have "aaaSbbbEccc" and want "aaaccc"? Perhaps `end` is actualy `length`?!
2. using the existing methods would make it easier for me to understand, though it is a secondary point